### PR TITLE
cpp: add build-time flags to make compression dependencies optional

### PR DESCRIPTION
**Pull request by james-rms**
### Public-Facing Changes

Adds two new build-time flags to optionally disable `zstd` and `lz4` support. These are `MCAP_COMPRESSION_NO_ZSTD` and `MCAP_COMPRESSION_NO_LZ4`.

When reading MCAP files, if the reader encounters a known compression method but is compiled without support for it, it will yield `Status::UnsupportedCompression`.

### Description

Based on https://github.com/foxglove/mcap/pull/1008

A new test binary is added to CI, which exercises these build-time flags.

### CODE REVIEW

File: cpp/mcap/include/mcap/errors.hpp [hunk 1]
```diff
@@ -23,6 +23,7 @@ enum class StatusCode {
   DecompressionFailed,
   DecompressionSizeMismatch,
   UnrecognizedCompression,
+  UnsupportedCompression,
```
Comment (jtbandes) [inline cpp/mcap/include/mcap/errors.hpp:26]: would there perhaps be a good reason to keep new status codes monotonically increasing (not modifying any existing status codes)? I'm not sure – just flagging as something to think about

File: cpp/test/unit_tests.cpp [hunk 2]
```diff
@@ -584,6 +584,7 @@ TEST_CASE("Message index records", "[writer]") {
   REQUIRE(messageIndexChannelIds[1] == channel2.id);
 }
 
+#ifndef MCAP_COMPRESSION_NO_LZ4
```
Comment (jtbandes) [inline cpp/test/unit_tests.cpp:587]: is there not a corresponding test case for zstd?

**Review by jtbandes — APPROVED** [review]
👍

File: cpp/mcap/include/mcap/errors.hpp (hunk 1 above)
Comment (james-rms) [reply cpp/mcap/include/mcap/errors.hpp:26]: Good idea.

File: cpp/test/unit_tests.cpp (hunk 2 above)
Comment (james-rms) [reply cpp/test/unit_tests.cpp:587]: There is not.

**Comment by asherikov** [comment]
Well, I was lazy with my PR, but if you are not I have a few suggestions:
- the defines should enable compression rather than disabling it;
- the defines should be located in corresponsing headers, so that the end users would need to include the headers they need instead of playing with defines;
- compression related stuff should be moved to mixin classes.

PS: oh, ok i guess too late.

**Pull request merged**